haskell/test/SMT/Types.hs

```diff
@@ -0,0 +1,141 @@
+-- SPDX-FileCopyrightText: 2021 TQ Tezos
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module SMT.Types
+  ( ModelError (..)
+  , contractErrorToModelError
+
+  , ContractCall (..)
+  , ContractCallSetting
+  , BaseDaoContractCall
+  , BlockchainState (..)
+  , ContractState (..)
+  , GeneratorConfig (..)
+  , GeneratorM
+  , PropertyTestInput (..)
+  , ActionType (..)
+
+  , toInt
+  , integerToNatural
+  ) where
+
+import Universum
+
+import Fmt
+import Hedgehog
+import Lorentz hiding (not, cast)
+import Tezos.Address (ContractHash(..))
+
+import Ligo.BaseDAO.Types
+
+-- The possible contract errors for the models
+data ModelError
+  = SOME_ERROR Text
+  | NOT_DELEGATE
+  | BAD_ENTRYPOINT_PARAMETER
+  | EMPTY_FLUSH
+  | MAX_PROPOSALS_REACHED
+  | BAD_STATE
+  | NOT_ENOUGH_FROZEN_TOKENS
+  | NOT_PROPOSING_STAGE
+  | PROPOSAL_NOT_UNIQUE
+  | PROPOSAL_NOT_EXIST
+  | EXPIRED_PROPOSAL
+  | MISSIGNED
+  | MAX_VOTERS_REACHED
+  | VOTING_STAGE_OVER
+  | DROP_PROPOSAL_CONDITION_NOT_MET
+  | NOT_ADMIN
+  | NOT_PENDING_ADMIN
+  deriving stock (Generic, Eq, Show)
+
+instance Buildable ModelError where
+  build = genericF
+
+instance (Buildable a, Buildable b) => Buildable (a, b) where
+  build  = genericF
+
+instance (Buildable a, Buildable b, Buildable c) => Buildable (a, b, c) where
+  build  = genericF
+
+contractErrorToModelError :: Text -> ModelError
+contractErrorToModelError txtError =
+ case txtError of
+   "NOT_DELEGATE" -> NOT_DELEGATE
+   "BAD_ENTRYPOINT_PARAMETER" -> BAD_ENTRYPOINT_PARAMETER
+   "EMPTY_FLUSH" -> EMPTY_FLUSH
+   "MAX_PROPOSALS_REACHED" -> MAX_PROPOSALS_REACHED
+   "BAD_STATE" -> BAD_STATE
+   "NOT_ENOUGH_FROZEN_TOKENS" -> NOT_ENOUGH_FROZEN_TOKENS
+   "NOT_PROPOSING_STAGE" -> NOT_PROPOSING_STAGE
+   "PROPOSAL_NOT_UNIQUE" -> PROPOSAL_NOT_UNIQUE
+   "PROPOSAL_NOT_EXIST" -> PROPOSAL_NOT_EXIST
+   "EXPIRED_PROPOSAL" -> EXPIRED_PROPOSAL
+   "MISSIGNED" -> MISSIGNED
+   "VOTING_STAGE_OVER" -> VOTING_STAGE_OVER
+   "DROP_PROPOSAL_CONDITION_NOT_MET" -> DROP_PROPOSAL_CONDITION_NOT_MET
+   "NOT_ADMIN" -> NOT_ADMIN
+   "NOT_PENDING_ADMIN" -> NOT_PENDING_ADMIN
+   _ -> SOME_ERROR txtError
+
+data ContractCall p = ContractCall
+  { ccSender :: Address
+  , ccAdvanceLevel :: Maybe Natural -- Advance level before calling the entrypoint
+  , ccChainId :: ChainId
+  , ccSelfAddress :: Address
+  , ccParameter :: p
+  , ccIdx :: Int
+  , ccGovernanceTokenAddress :: ContractHash
+  , ccSource :: Address
+  } deriving stock (Show, Generic)
```

Comment:
This is also kinda odd to me:

Shouln't `ccChainId`, `ccSelfAddress`, `ccGovernanceTokenAddress` be moved to the `ModelState`?

Moreover, I think that `ccIdx` is unnecessary if we do the step-by-step SMT checks as said above.

Reply:
I've removed the `cIdx`, and includes those type in `ModelState`.